src/pudl/transform/eia861.py

```diff
@@ -175,14 +434,243 @@
+    )
+    return deduped_df
+
+
+def _check_for_dupes(df, df_name, subset):
+    dupes = (
+        df.duplicated(
+            subset=subset, keep=False)
+    )
+    if dupes.any():
+        raise AssertionError(
+            f"Found {len(dupes)} duplicate rows in the {df_name} table, "
+            f"when zero were expected!"
+        )
+
+
+def _early_transform(df):
+    """Fix EIA na values and convert year column to date."""
+    df = pudl.helpers.fix_eia_na(df)
+    df = pudl.helpers.convert_to_date(df)
+    return df
+
+
+def _compare_totals(data_cols, idx_cols, class_type, df_name):
+    """Compare reported totals with sum of component columns.
+
+    Args:
+        data_cols (pd.DataFrame): A DataFrame containing only the columns with
+            normalized information.
+        idx_cols (list): A list of the primary keys for the given denormalized
+            DataFrame.
+        class_type (str): The name (either 'customer_class' or 'tech_class') of
+            the column for which you'd like to compare totals to components.
+        df_name (str): The name of the dataframe.
+    """
+    # Convert column dtypes so that numeric cols can be adequately summed
+    data_cols = pudl.helpers.convert_cols_dtypes(data_cols, 'eia')
+    # Drop data cols that are non numeric (preserve primary keys)
+    data_cols = (
+        data_cols.set_index(idx_cols + [class_type])
+        .select_dtypes('float', 'Int64')
+        .reset_index()
+    )
+    # Create list of data columns to be summed (may include non-numeric that will be excluded)
+    data_col_list = set(data_cols.columns.tolist()) - \
+        set(idx_cols + [class_type])
+    # Distinguish reported totals from segments
+    data_totals_df = data_cols.loc[data_cols[class_type] == 'total']
+    data_no_tots_df = data_cols.loc[data_cols[class_type] != 'total']
+    # Calculate sum of segments for comparison against reported total
+    data_sums_df = data_no_tots_df.groupby(
+        idx_cols + [class_type], observed=True).sum()
+    sum_total_df = pd.merge(data_totals_df, data_sums_df, on=idx_cols,
+                            how='outer', suffixes=('_total', '_sum'))
+    # Check each data column's calculated sum against the reported total
+    for col in data_col_list:
+        col_df = (sum_total_df.loc[sum_total_df[col + '_total'].notnull()])
+        if len(col_df) > 0:
+            col_df = (
+                col_df.assign(
+                    compare_totals=lambda x: (x[col + '_total'] == x[col + '_sum']))
+            )
+            pct_bad_math = str(int(
+                (~col_df['compare_totals']).values.sum()
+                / len(col_df)
+                * 100
+            )) + '%'
+            log_output = f'{df_name}: for column {col}, {pct_bad_math} of non-null reported totals ≠ the sum of parts.'
+        else:
+            log_output = f'{df_name}: for column {col} all total values are nan'
+        print(log_output)
```

Comment:
Oh! this is from your last PR, but this `print` should probably be a `logger.info`

Reply:
Oh, yes! Thanks for pointing that out - I changed it to print because I have the loggers temporarily turned off (too many print lines)

Reply:
You can change how much output the loggers generate by using e.g. `logger.setLevel("WARNING")` (less output) or `logger.setLevel("DEBUG")` (more output).

Reply:
Oh neat! Is that something I'd set in the notebook??

Reply:
You can set it in the notebook when you create the `logger` instance, or you can set it temporarily and then re-set it after something that makes too much output.  I've also felt like some of the extract process is just putting out too much stuff all at the `INFO` level -- I will change some of the loglevels to `DEBUG` and others to `WARNING` so that there's more control in there. Overall there are 5 loglevels, in order: `DEBUG`, `INFO`, `WARNING`, `ERROR`, and `CRITICAL` and the idea is to be able to adjust how verbose the output is based on your context.

Reply:
The logging library is super flexible and I definitely don't understand how to use all of it yet. There are a bunch of good tutorials out there, this is one overview: https://realpython.com/python-logging/